### PR TITLE
Lock screens

### PR DESCRIPTION
**Pull request by ptomato**
https://phabricator.endlessm.com/T23929

### CODE REVIEW

File: src/lockscreen.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/* exported Lockscreen */
+
+const {GObject, Gtk} = imports.gi;
+
+const _css = `
+.lockscreen {
+    background: url('/usr/share/icons/Adwaita/256x256/status/dialog-password.png'),
```
Comment (cosimoc) [inline src/lockscreen.js]: I know this is a stub, but would `-gtk-icontheme` work instead of hardcoding a path?

File: src/lockscreen.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        this._updateUI();
+        this.notify('locked');
+    }
+
+    _updateUI() {
+        if (!this._overlay)
+            return;
+
+        const style = this._overlay.get_style_context();
+        style.add_class('lockscreen');
```
Comment (cosimoc) [inline src/lockscreen.js]: This should not be required, the style class is already always added in `_init()`

File: src/lockscreen.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        super._init(props);
+
+        this._overlay = new Gtk.Frame({
+            expand: true,
+            visible: true,
+        });
+        this.add_overlay(this._overlay);
+
+        const style = this._overlay.get_style_context();
+        style.add_class('lockscreen');
```
Comment (cosimoc) [inline src/lockscreen.js]: Instead of a style class, you should be able to set a CSS name for this class; gjs has a way to do that directly as part of the class definition.

File: src/window.js [hunk 4]
```diff
@@ -64,6 +72,23 @@ var ToolboxWindow = GObject.registerClass({
...
+
+    unlock() {
+        if (!(this.target_bus_name in unlockState))
+            unlockState[this.target_bus_name] = [true];
+        const index = unlockState[this.target_bus_name].findIndex(state => !state);
+        unlockState[this.target_bus_name][index] = true;
+        this.emit('unlock-state-changed');
+
+        // Update state of the common lockscreen in this window
+        this._lockscreen.locked = !this.getUnlockState()[0];
```
Comment (cosimoc) [inline src/window.js:105]: This belongs to a different commit

File: src/window.js [hunk 5]
```diff
@@ -70,6 +70,12 @@ var ToolboxWindow = GObject.registerClass({
         this._lockscreen.add(this._toolbox_frame);
 
         this.get_style_context().add_class('toolbox-surrounding-window');
+
+        const accelGroup = new Gtk.AccelGroup();
+        accelGroup.connect('r'.charCodeAt(),
+            Gdk.ModifierType.SHIFT_MASK | Gdk.ModifierType.CONTROL_MASK, 0,
+            this._onKeyCombo.bind(this));
+        this.add_accel_group(accelGroup);
```
Comment (cosimoc) [inline src/window.js]: It's probably better to make this a GAction in the window action group, in case we ever need to unlock this remotely (and it's more natural to add a keyboard shortcut for it)

**Review by cosimoc — CHANGES_REQUESTED** [review]

File: src/lockscreen.js (hunk 1 above)
Comment (ptomato) [reply src/lockscreen.js]: I don't think so, as we don't call `gtk_render_icon()` anywhere in this widget.

File: src/lockscreen.js (hunk 3 above)
Comment (ptomato) [reply src/lockscreen.js]: I want the style class on the overlay widget, instead of the class itself. However, since the style provider only applies to the one widget anyway, I'll just use its existing CSS name (`frame`).

File: src/window.js (hunk 4 above)
Comment (ptomato) [reply src/window.js:105]: Fixed.

**Comment by ptomato** [comment]
Should be all fixed now.
